library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ModNioResourcePack.java

```diff
@@ -0,0 +1,215 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.resource.loader.impl;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.metadata.ModMetadata;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.resource.AbstractFileResourcePack;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.InvalidIdentifierException;
+
+import org.quiltmc.qsl.resource.loader.api.ResourcePackActivationType;
+
+/**
+ * A NIO implementation of a mod resource pack.
+ */
+public class ModNioResourcePack extends AbstractFileResourcePack {
+	private static final Logger LOGGER = LogManager.getLogger();
+	private static final Pattern RESOURCE_PACK_PATH = Pattern.compile("[a-z0-9-_]+");
+	private final ModMetadata modInfo;
+	private final Path basePath;
+	private final ResourceType type;
+	private final boolean cacheable;
+	private final AutoCloseable closer;
+	private final String separator;
+	private final ResourcePackActivationType activationType;
+	private Set<String> namespaceCache;
+
+	public ModNioResourcePack(ModMetadata modInfo, Path path, ResourceType type, AutoCloseable closer,
+							  ResourcePackActivationType activationType) {
+		super(null);
+		this.modInfo = modInfo;
+		this.basePath = path.toAbsolutePath().normalize();
+		this.type = type;
+		// Cache ModNioResourcePacks if not in dev environment because it's not supposed to be mutable in production.
+		this.cacheable = modInfo.getId().equals("minecraft") || !FabricLoader.getInstance().isDevelopmentEnvironment();
+		this.closer = closer;
+		this.separator = basePath.getFileSystem().getSeparator();
+		this.activationType = activationType;
+	}
+
+	private Path getPath(String filename) {
+		Path childPath = this.basePath.resolve(filename.replace("/", separator)).toAbsolutePath().normalize();
+
+		if (childPath.startsWith(basePath) && Files.exists(childPath)) {
+			return childPath;
+		} else {
+			return null;
+		}
+	}
+
+	@Override
+	protected InputStream openFile(String filename) throws IOException {
+		InputStream stream;
+
+		Path path = this.getPath(filename);
+
+		if (path != null && Files.isRegularFile(path)) {
+			return Files.newInputStream(path);
+		}
+
+		stream = ModResourcePackUtil.openDefault(this.modInfo, this.type, filename);
+
+		if (stream != null) {
+			return stream;
+		}
+
+		// ReloadableResourceManagerImpl gets away with FileNotFoundException.
```

Comment:
gets away? what does that mean?

Reply:
To me it reads like "resource not found, so fail" in surrounding code.

Reply:
It means that here the resource manager will catch the exception and return `null` iirc.